src/ServiceControl.Monitoring/Http/Diagrams/DiagramApiModule.cs

```diff
@@ -15,78 +15,118 @@ public class MonitoredEndpointsModule : ApiModule
+                CreateMetric<EndpointMessageType, ProcessingTimeStore>("Throughput", IntervalsAggregator.AggregateTotalMeasurementsPerSecond),
+            };
+
             Get["/monitored-endpoints"] = parameters =>
             {
                 var endpoints = GetMonitoredEndpoints(endpointRegistry, activityTracker);
                 var period = ExtractHistoryPeriod();
-               
-                FillInEndpointData(endpoints, criticalTimeStore, period, (e, v) => e.CriticalTime = v, IntervalsAggregator.AggregateTimings);
-                FillInEndpointData(endpoints, processingTimeStore, period, (e, v) => e.ProcessingTime = v, IntervalsAggregator.AggregateTimings);
-                FillInEndpointData(endpoints, retriesStore, period, (e, v) => e.Retries = v, IntervalsAggregator.AggregateRetries);
-                FillInEndpointData(endpoints, queueLengthStore, period, (e, v) => e.QueueLength = v, IntervalsAggregator.AggregateQueueLength);
-                FillInEndpointData(endpoints, processingTimeStore, period, (e, v) => e.Throughput = v, intervals => IntervalsAggregator.AggregateTotalMeasurementsPerSecond (intervals, period));
+
+                foreach (var metric in instanceMetric)
+                {
+                    var store = metricByInstance.Single(s => s.GetType() == metric.StoreType);
+                    var intervals = store.GetIntervals(period, DateTime.UtcNow).ToLookup(k => k.Id.EndpointName);
+
+                    foreach (var endpoint in endpoints)
+                    {
+                        var values = metric.Aggregate(intervals[endpoint.Name].ToList(), period);
+
+                        endpoint.Metrics.Add(metric.ReturnName, values);
+                    }
+                }
 
                 return Negotiate.WithModel(endpoints);
             };
 
             Get["/monitored-endpoints/{endpointName}"] = parameters =>
             {
                 var endpointName = (string)parameters.EndpointName;
+                var period = ExtractHistoryPeriod();
 
                 var instances = GetMonitoredEndpointInstances(endpointRegistry, endpointName, activityTracker);
-                var period = ExtractHistoryPeriod();
 
-                FillInInstanceData(instances, criticalTimeStore, period, (e, v) => e.CriticalTime = v, IntervalsAggregator.AggregateTimings);
-                FillInInstanceData(instances, processingTimeStore, period, (e, v) => e.ProcessingTime = v, IntervalsAggregator.AggregateTimings);
-                FillInInstanceData(instances, retriesStore, period, (e, v) => e.Retries = v, IntervalsAggregator.AggregateRetries);
-                FillInInstanceData(instances, processingTimeStore, period, (e, v) => e.Throughput = v, intervals => IntervalsAggregator.AggregateTotalMeasurementsPerSecond(intervals, period));
+                foreach (var metric in instanceMetric)
+                {
+                    var store = metricByInstance.Single(s => s.GetType() == metric.StoreType);
```

Comment:
This should be a dictionary initialized as a field in ctor and lookup should be done by indexer [metric.StoreType]